src/main/groovy/org/cadixdev/gradle/licenser/LicenseExtension.groovy

```diff
@@ -103,16 +104,39 @@ class LicenseExtension extends LicenseProperties {
     @Incubating
     final NamedDomainObjectContainer<LicenseTaskProperties> tasks
 
-    LicenseExtension(Project project) {
-        super(new PatternSet())
+    @PackageScope final ObjectFactory objects
+    @PackageScope final TextResourceFactory textResources
+    @PackageScope final ProviderFactory providers
 
-        this.tasks = project.container(LicenseTaskProperties) { name ->
-            new LicenseTaskProperties((filter as PatternSet).intersect(), name)
+    @Inject
+    LicenseExtension(final ObjectFactory objects, final Project project) {
+        super(new PatternSet(), objects, project.resources.text)
+
+        this.objects = objects
+        this.textResources = project.resources.text
+        this.providers = project.providers
+        this.tasks = objects.domainObjectContainer(LicenseTaskProperties) { String name ->
+            new LicenseTaskProperties((filter as PatternSet).intersect(), name, objects, textResources, charset)
         }
+        this.conditionalProperties = objects.listProperty(LicenseProperties)
 
         // Defaults
-        newLine = true
-        charset = 'UTF-8'
+        this.keywords = objects.listProperty(String).convention(['Copyright', 'License'])
+        this.ignoreFailures = objects.property(Boolean).convention(false)
+        this.skipExistingHeaders = objects.property(Boolean).convention(false)
+        this.charset = objects.property(String).convention('UTF-8')
+        super.charset.set(this.charset)
+
+        def defaultLineEnding
+        try {
+            // Gradle 6+
+            defaultLineEnding = project.providers.systemProperty("line.separator")
+        } catch (final MissingMethodException ex) {
+            // Gradle 5.x TODO @ 0.7: remove this
+            defaultLineEnding = project.provider { System.lineSeparator() }
+        }
+        this.lineEnding = objects.property(String).convention(defaultLineEnding)
+        this.newLine.convention(true)
```

Comment:
Couldn't we set the convention for the `newLine` property only in the `LicenseProperties` constructor?

Reply:
The property setup is designed so that `LicenseProperties` instances can inherit their values from the extension, which means only the property in the extension should have a default -- null everywhere else means "delegate to extension"